File: nattybear/level2/기능개발/solution.js
function solution(progresses, speeds) {
  return group(zip_with(days, progresses, speeds))
    .map(x => x.length)
}

function days(progress, speed) {
  todo = 100 - progress
  quotient = Math.floor(todo / speed)
  remainder = todo % speed
  if (remainder == 0)
    return quotient
  else
    return quotient + 1
}

function zip_with(f, xs, ys) {
  const zs = []
  for (let i = 0; i < xs.length; i++) {
    let x = xs[i]
    let y = ys[i]
    let z = f(x, y)
    zs.push(z)
  }
  return zs
}

function group(xs) {
  const yys = []
  let ys = []
  for (let i = 0; i < xs.length; i++) {
    let x = xs[i]
    if (ys.length == 0) {
      ys.push(x)
      continue
    }
    if (ys[0] >= x) {
      ys.push(x)
    }
    else {
      yys.push(ys)
      ys = [x]
    }
  }
  yys.push(ys)
  return yys
}

File: nattybear/level2/기능개발/Solution.java
import java.util.ArrayList;

class Ints extends ArrayList<Integer> {}

class Solution {
  public int[] solution(int[] progresses, int[] speeds) {
    return group(zip(progresses, speeds));
  }

  static int[] zip(int[] progresses, int[] speeds) {
    int size = speeds.length;
    int[] xs = new int[size];
    for (int i = 0; i < size; i++) {
      int p = progresses[i];
      int s = speeds[i];
      int x = days(p, s);
      xs[i] = x;
    }
    return xs;
  }

  static int days(int progress, int speed) {
    int todo = 100 - progress;
    int quotient = todo / speed;
    int remainder = todo % speed;
    if (remainder == 0)
      return quotient;
    else
      return quotient + 1;
  }

  static int[] group(int[] xs) {
    Ints zs = new Ints();
    Ints ys = new Ints();
    for (int i = 0; i < xs.length; i++) {
      int x = xs[i];
      if (ys.size() == 0) {
        ys.add(x);
        continue;
      }
      if (ys.get(0) >= x) {
        ys.add(x);
      }
      else {
        zs.add(ys.size());
        ys.clear();
        ys.add(x);
      }
    }
    zs.add(ys.size());
    return zs.stream().mapToInt(x -> x).toArray();
  }
}
